Userland/Demos/CatDog/CatDog.cpp

```diff
@@ -10,174 +10,187 @@
-    } else {
-        m_left = false;
-        m_right = false;
-        m_moveX = 0;
-    }
 
-    if (m_temp_pos.y() > 48) {
-        m_up = false;
-        m_down = true;
-        m_moveY = 10;
-
-        m_curr_bmp = m_srun1;
-        if (m_curr_frame == 2)
-            m_curr_bmp = m_srun2;
-    } else if (m_temp_pos.y() < -16) {
-        m_up = true;
-        m_down = false;
-        m_moveY = -10;
-
-        m_curr_bmp = m_nrun1;
-        if (m_curr_frame == 2)
-            m_curr_bmp = m_nrun2;
-    } else {
-        m_up = false;
-        m_down = false;
-        m_moveY = 0;
-    }
+    Gfx::IntPoint move;
+    auto const size = window()->size();
 
-    if (m_up && m_left) {
-        m_curr_bmp = m_nwrun1;
-        if (m_curr_frame == 2)
-            m_curr_bmp = m_nwrun2;
-    } else if (m_up && m_right) {
-        m_curr_bmp = m_nerun1;
-        if (m_curr_frame == 2)
-            m_curr_bmp = m_nerun2;
-    } else if (m_down && m_left) {
-        m_curr_bmp = m_swrun1;
-        if (m_curr_frame == 2)
-            m_curr_bmp = m_swrun2;
-    } else if (m_down && m_right) {
-        m_curr_bmp = m_serun1;
-        if (m_curr_frame == 2)
-            m_curr_bmp = m_serun2;
+    if (m_mouse_offset.x() < 0) {
+        m_state |= State::LEFT;
+        move.set_x(max(m_mouse_offset.x(), -size.width() / 2));
+    } else if (m_mouse_offset.x() > size.width()) {
+        m_state |= State::RIGHT;
+        move.set_x(min(m_mouse_offset.x(), size.width() / 2));
     }
 
-    window()->move_to(window()->position().x() + m_moveX, window()->position().y() + m_moveY);
-    m_temp_pos.set_x(m_temp_pos.x() + (-m_moveX));
-    m_temp_pos.set_y(m_temp_pos.y() + (-m_moveY));
+    if (m_mouse_offset.y() < 0) {
+        m_state |= State::UP;
+        move.set_y(max(m_mouse_offset.y(), -size.height() / 2));
+    } else if (m_mouse_offset.y() > size.height()) {
+        m_state |= State::DOWN;
+        move.set_y(min(m_mouse_offset.y(), size.height() / 2));
+    }
 
-    if (m_curr_frame == 1) {
-        m_curr_frame = 2;
+    if ((m_state & DIRECTION_MASK) == 0) {
+        if (m_idle_sleep_timer.elapsed() > 5'000)
+            m_state |= State::SLEEPING;
     } else {
-        m_curr_frame = 1;
+        m_state &= ~State::IDLE;
+        m_idle_sleep_timer.start();
     }
 
-    if (!m_up && !m_down && !m_left && !m_right) {
-        // Select the movement-free image based on the main state.
-        if (m_timer.elapsed() > 5000)
-            m_main_state = MainState::Sleeping;
-        set_image_by_main_state();
-    } else if (is_non_application_state(m_main_state)) {
-        // If CatDog currently moves, it should be idle the next time it stops.
-        m_main_state = MainState::Idle;
-    }
+    window()->move_to(window()->position() + move);
+    m_mouse_offset -= move;
+
+    // Toggle FRAME_2 bit every other frame
+    m_state |= (++m_frame & State::FRAME_2);
 
     update();
 }
 
+RefPtr<Gfx::Bitmap> CatDog::bitmap_for_state() const
```

Comment:
Does this need to incref every frame? Returning a `Gfx::Bitmap&` could reduce that churn if we know that the images vector will never be modified after init.

Reply:
Fixed.